### ISSUE TITLE
Question: How to adapt MT plugin for memoQ Server 9.0 integration?

### ISSUE THREAD

**Issue by PoliakovaAN** [JuchiaLu/Multi-Supplier-MT-Plugin#30]
Dear LuchiaLu,

Thank you very much for your work on memoQ plugins — it has been truly helpful.

My name is Irina. I am a linguist and work at a leading medical translation agency in Moscow, where we use memoQ (server version 9.0). At the same time, I am studying NLP at the Higher School of Economics. Recently, I fine-tuned a local NMT model on six million high-quality bilingual segments collected by our agency. Integrated through FastAPI (which I wrote myself with ChatGPT’s help), the model delivers excellent results — often outperforming Google Translate — and works perfectly in memoQ 9.14 on standalone projects.

The problem is that the plugin does not appear when I try to translate local copies of server projects. This makes it impossible to use my model in our production workflow, since memoQ 9.14 must remain compatible with our 9.0 server.

That is why I would like to develop a similar plugin for memoQ Server 9.0. From what I understand, your plugins are mainly for local versions, but I wanted to ask: do you have any advice, documentation, or example code related to server-side integration? I have a developer colleague who can assist me, and any guidance you could share would be greatly appreciated.

Thank you again for your work and for any support you might be able to provide!

Best regards,
Irina

<img width="481" height="663" alt="Image" src="https://github.com/user-attachments/assets/4d7312c9-b0c8-4929-9ec7-1bc8749aa561" />

**Comment by JuchiaLu** [comment]
Hi Irina,

You’re right — the latest release of the plugin (v1.4.0) isn’t fully compatible with memoQ Server. Someone else reported the [same issue](https://github.com/JuchiaLu/Multi-Supplier-MT-Plugin/issues/27) earlier, and I shared a [temporary patched version (v1.4.0.1)](https://github.com/user-attachments/files/21145928/Multi-Supplier-MT-Plugin-1.4.0.1.zip). According to them, it works on memoQ Server, but the configuration settings might not be saved correctly. Since I don’t have memoQ Server installed myself, I can’t confirm whether that limitation is real.

From a development perspective, there’s no real difference between local-side and server-side plugins — the code is the same. The difference is in how they run: on the server, the plugin and its configuration are stored centrally and delivered to memoQ Client over the network. The client doesn’t keep the DLLs or settings locally, it just loads them from the server each time.

A quick question: does your FastAPI wrapper expose an OpenAI-compatible API? If yes, you can probably try the temporary build I mentioned. If not, are you thinking of building a plugin from scratch, or would you rather adapt the existing plugin to talk to your FastAPI service?

If you want to build one from scratch, you can grab the [Machine Translation SDK](https://www.memoq.com/integrations/sdks/) from the memoQ website — the ZIP includes documentation (in Word format) that explains how to develop plugins. If instead you’d like to adapt this plugin to your FastAPI, there are two options:

- You can share the API documentation for your FastAPI service, and I can help integrate it directly.

- Or, if you’d prefer to do it yourself, let me know and I’ll put together a short guide on how to extend the current plugin to support a new provider.

**Comment by PoliakovaAN** [comment]
Hi LuchiaLu,

Thank you again for the detailed explanation!

Here is the core of my FastAPI service (based on a fine-tuned MarianMT model):

@app.post("/translate")
async def translate_text(req: Request):
    data = await req.json()

    if "messages" in data:
        user_messages = [m["content"] for m in data.get("messages", []) if m["role"] == "user"]
        text = user_messages[-1] if user_messages else ""
    elif "source" in data:
        text = data["source"]
    else:
        return {"translation": ""}

    inputs = tokenizer([text], return_tensors="pt", padding=True)
    with torch.no_grad():
        output = model.generate(**inputs)
    translation = tokenizer.decode(output[0], skip_special_tokens=True)

    # minimal OpenAI-compatible response
    return {
        "id": "test",
        "object": "chat.completion",
        "choices": [
            {
                "index": 0,
                "message": {"role": "assistant", "content": translation},
                "finish_reason": "stop"
            }
        ]
    }

So the service already returns a minimal OpenAI-compatible response (chat.completion with choices → message → content).
Do you think this will be enough for the patched build (v1.4.0.1) to work with memoQ Server projects?

Also, since I am not a system administrator myself, could you please suggest what instructions or key points I should pass on to our IT admins to make the deployment of this plugin on memoQ Server smoother? For example: where the plugin DLLs need to be placed, what network configuration may be required, and whether they need to adjust any memoQ Server settings.

Thanks a lot for your help and for any practical guidance you can share!

Best regards,
Irina

[server.py](https://github.com/user-attachments/files/21897387/server.py)

**Comment by JuchiaLu** [comment]
I may have misunderstood you earlier. I thought you wanted to integrate your FastAPI service directly into this plugin (Multi-Supplier-MT-Plugin). But when you wrote *“and works perfectly in memoQ 9.14 on standalone projects”*, it sounds like you already have your own plugin running fine in memoQ 9.14, and what you’re asking is how to make **your plugin** work with memoQ Server — is that correct?

If that’s the case, you might find these memoQ help pages useful:

* [Setting up machine translation plugins on memoQ TMS](https://docs.memoq.com/helpcenter/memoQ-TMS/Setting-up-machine-translation-plugins-on-memoQ-TMS.htm)
* [View and manage online MT settings (on memoQ TMS)](https://docs.memoq.com/current/en/Workspace/resconsole-mtsettings.html)

Since I’m not a translator and haven’t worked with memoQ Server myself, I’m not very familiar with its setup — so unfortunately I can’t give you much hands-on advice here.

On the other hand, if you really do want to integrate your FastAPI with this plugin (Multi-Supplier-MT-Plugin), then let me clarify something I missed before: I originally assumed your model was an LLM, but now I see it’s an NMT system. That’s why I mistakenly suggested exposing an OpenAI-compatible API. In practice, OpenAI-style APIs aren’t a great fit for NMT, since they don’t support passing language codes.

A better approach would be to make your FastAPI expose a **DeepL-compatible API** (see [DeepL API reference](https://developers.deepl.com/api-reference/translate/request-translation)). That way, you can hook up your translation service using the DeepL configuration in this plugin without extra custom work.

**Comment by PoliakovaAN** [comment]
Thanks a lot for your reply and the links! I used an OpenAI-compatible API because it was the only format I could get working, and it seems to run fine locally on EN→RU. Thanks also for the suggestion about using a DeepL-compatible API — I’ll try it and let you know how it goes.

**Comment by PoliakovaAN** [comment]
Hi, JuchiaLu

One more question — would you recommend using the Multi-Supplier-MT-Plugin v1.4.0.1 with memoQ Server 9.0.24? I’d just like to double-check, since our memoQ Server is quite an old version.

<img width="615" height="194" alt="Image" src="https://github.com/user-attachments/assets/00093a20-06dc-4a07-944b-99c3d12c3fd9" />

**Comment by JuchiaLu** [comment]
In general, plugins don’t actually run on the Server itself — they’re delivered to the Client and executed there. Since your memoQ Client is version 9.14, it should normally work, because that’s the minimum version officially supported by the plugin.

The only uncertainty comes from your Server being 9.0.24, which was released earlier than the Client. That’s why I can’t guarantee 100% compatibility.

For a fallback, I previously built a version specifically for memoQ releases earlier than 9.14 (see [issue](https://github.com/JuchiaLu/Multi-Supplier-MT-Plugin/issues/29) and the download link: [Multi-Supplier-MT-Plugin-1.4.0.3.zip](https://github.com/user-attachments/files/21454601/Multi-Supplier-MT-Plugin-1.4.0.3.zip)). This build doesn’t support advanced features like context, summaries, or TM integration — but since your service is NMT, those aren’t really needed anyway.

So, in short: I’d say there’s about a 90% chance it will run fine as is, but if not, you can fall back to version 1.4.0.3.

**Comment by IrenePolyakova** [comment]
Dear LuchiaLu,

This is Irina, I’m just writing from another account, and I’d like to ask a follow-up question related to this topic.
My colleague is currently working on a plugin for memoQ Server 9.0.4, using your code as a base. He sent me a question for you:

Could you please help us understand what might be causing the following issue? When using the plugin on the server, the logs show this warning:
2025.09.28 12:04:29.0547172 memoQServer (Warning)  () The current MT plugin is installed, but it is not configured properly for the eng-rus language pair (or its entire configuration is missing).

I slightly modified the plugin according to https://docs.memoq.com/9-12/sdk-docs/memoQ-MT-SDK-2.1.3.zip
Replaced ISessionWithMetadata with ISession

Removed MTRequestMetadata

From the logs, it looks like the CreateLookupSession() method is not even being called.

The full code is here: 
https://github.com/at146/Multi-Supplier-MT-Plugin

Environment:

memoQ Server 9.0.24 (Windows Server 2012 R2, Microsoft SQL Server 2012)

memoQ Client 9.14 (Windows 11 Pro)

Could you kindly take a look and let us know what might be wrong?

With respect,
Irina

**Comment by JuchiaLu** [comment]
Have you tried running either of these builds — [Multi-Supplier-MT-Plugin-1.4.0.1.zip](https://github.com/user-attachments/files/21145928/Multi-Supplier-MT-Plugin-1.4.0.1.zip) or [Multi-Supplier-MT-Plugin-1.4.0.3.zip](https://github.com/user-attachments/files/21454601/Multi-Supplier-MT-Plugin-1.4.0.3.zip) — on memoQ Server 9.0.24? Do they work properly for you?

These two builds include some adjustments specifically for memoQ Server, but the changes weren’t committed to the repository. If they run fine on your end, I can share the source code for these versions so you’ll have a proper base to work from.

**Comment by IrenePolyakova** [comment]
Dear LuchiaLu, 
neither of these plugin versions worked. 
My colleague-developer, Alexey, reported that to make the plugin work, he copied and replaced the files from the server folder (Kilgray.Utils.dll, MemoQ.Addins.Common.dll, MemoQ.MTInterfaces.dll, Newtonsoft.Json.dll) into the References folder, and after that, everything worked. The problem was with the Newtonsoft.Json package, as the package version on the server was older than the one on the client (13.0.3).
A huge thank you for your source code. The plugin can now run on the bureau's server with my model and effectively translate complex, domain-specific texts.

**Comment by mohu113** [comment]
> The plugin can now run on the bureau's server with my model and effectively translate complex, domain-specific texts.

Kindly share your modified server version

Thanks in advance

**Comment by IrenePolyakova** [comment]
Hello, 
Here is the link from Alexey: 
https://github.com/at146/Multi-Supplier-MT-Plugin/tree/server-9.0.24.24340
